libyul/optimiser/VarNameCleaner.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/VarNameCleaner.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <algorithm>
+#include <cctype>
+#include <climits>
+#include <iterator>
+#include <string>
+#include <regex>
+
+using namespace yul;
+using namespace std;
+
+void VarNameCleaner::operator()(FunctionDefinition& _funDef)
+{
+	m_usedNames = m_blacklist;
+	m_translatedNames.clear();
+
+	ASTModifier::operator()(_funDef);
+}
+
+void VarNameCleaner::operator()(VariableDeclaration& _varDecl)
+{
+	for (TypedName& typedName: _varDecl.variables)
+		if (auto newName = makeCleanName(typedName.name))
+			typedName.name = *newName;
+
+	ASTModifier::operator()(_varDecl);
+}
+
+void VarNameCleaner::operator()(Identifier& _identifier)
+{
+	if (auto newName = newlyAssignedName(_identifier.name))
+		_identifier.name = *newName;
+}
+
+boost::optional<YulString> VarNameCleaner::makeCleanName(YulString const& _name)
+{
+	if (auto newName = findCleanName(_name))
+	{
+		m_usedNames.insert(*newName);
+		m_translatedNames[_name] = *newName;
+
+		return newName;
+	}
+
+	// Name isn't used yet, but we need to make sure nobody else does.
+	m_usedNames.insert(_name);
+	return boost::none;
+}
+
+bool VarNameCleaner::isUsedName(YulString const& _name) const
+{
+	return m_usedNames.find(_name) != m_usedNames.end();
+}
+
+boost::optional<YulString> VarNameCleaner::findCleanName(YulString const& _name) const
+{
+	if (auto newName = stripSuffix(_name))
+	{
+		if (newName->str().length() != 0 && !m_dialect.builtin(*newName) && !m_blacklist.count(*newName) && !m_usedNames.count(*newName))
+			return newName;
+
+		// create new name with suffix (by finding a free identifier)
+		for (size_t i = 1; i < numeric_limits<decltype(i)>::max(); ++i)
+		{
+			YulString newNameSuffixed = YulString{newName->str() + "_" + to_string(i)};
+			if (!isUsedName(newNameSuffixed) && !m_blacklist.count(newNameSuffixed))
```

Comment:
Some here: ``!isUsedName(newNameSuffixed)`` should already imply ``!m_blacklist.count(newNameSuffixed)``, since ``m_blacklist`` is a usbset of ``m_usedNames``, right?